stl/inc/ranges

```diff
@@ -4749,6 +4752,474 @@ namespace ranges {
+                    --_Outer_it;
+                    auto&& _Inner = *_Outer_it;
+                    _Inner_it._Emplace_second(_RANGES end(_Inner));
+                }
+
+                while (true) {
+                    if (_Inner_it._Contains == _Variantish_state::_Holds_first) {
+                        auto& _It = _Inner_it._First;
+                        if (_It == _RANGES begin(_Parent->_Pattern)) {
+                            --_Outer_it;
+                            auto&& _Inner = *_Outer_it;
+                            _Inner_it._Emplace_second(_RANGES end(_Inner));
+                        } else {
+                            break;
+                        }
+                    } else if (_Inner_it._Contains == _Variantish_state::_Holds_second) {
+                        auto& _It     = _Inner_it._Second;
+                        auto&& _Inner = *_Outer_it;
+                        if (_It == _RANGES begin(_Inner)) {
+                            _Inner_it._Emplace_first(_RANGES end(_Parent->_Pattern));
+                        } else {
+                            break;
+                        }
+                    } else {
+                        _Throw_bad_variant_access();
+                    }
+                }
+
+                switch (_Inner_it._Contains) {
+                case _Variantish_state::_Holds_first:
+                    --_Inner_it._First;
+                    break;
+                case _Variantish_state::_Holds_second:
+                    --_Inner_it._Second;
+                    break;
+                case _Variantish_state::_Nothing:
+                    _Throw_bad_variant_access();
+                }
+                return *this;
+            }
+
+            constexpr _Iterator operator--(int) requires _Deref_is_glvalue && bidirectional_range<_Base> //
+                && _Bidirectional_common<_InnerRng<_Const>> && _Bidirectional_common<_PatternBase> {
+                auto _Tmp = *this;
+                --*this;
+                return _Tmp;
+            }
+
+            friend constexpr bool operator==(const _Iterator& _Left, const _Iterator& _Right) requires _Deref_is_glvalue
+                && equality_comparable<_OuterIter> && equality_comparable<_InnerIter> {
+                if (_Left._Outer_it == _Right._Outer_it) {
+                    if (_Left._Inner_it._Contains == _Right._Inner_it._Contains) {
+                        switch (_Left._Inner_it._Contains) {
+                        case _Variantish_state::_Holds_first:
+                            return _Left._Inner_it._First == _Right._Inner_it._First;
+                        case _Variantish_state::_Holds_second:
+                            return _Left._Inner_it._Second == _Right._Inner_it._Second;
+                        case _Variantish_state::_Nothing:
+                            return true;
+                        }
+                    }
+                }
+                return false;

```

Comment:
Nitpick: I would prefer early returns to nesting but this is more or less cosmetic
```suggestion
                if (_Left._Outer_it != _Right._Outer_it) {
                    return false;
                }
                
                if (_Left._Inner_it._Contains != _Right._Inner_it._Contains) {
                    return false;
                }
                 
                switch (_Left._Inner_it._Contains) {
                case _Variantish_state::_Holds_first:
                    return _Left._Inner_it._First == _Right._Inner_it._First;
                case _Variantish_state::_Holds_second:
                    return _Left._Inner_it._Second == _Right._Inner_it._Second;
                case _Variantish_state::_Nothing:
                    return true;
                }
```